### PR TITLE
Ng meta merge

### PR DESCRIPTION
**Pull request by kegluneq**
These commits collectively merge the `DirectiveMetadataLinker` and `DirectiveLinker` phases of the transformer, which should avoid a lot of extra asset requests.

This also gets rid of `EmptyNgDepsRemover`, which only affects output code size, not correctness. It will be completely obsolete when we update to use const, tree-shakable maps, so I'm removing it now for simplicity.

Blocked by #3605

### CODE REVIEW

**Comment by yjbanov** [comment]
Do you mean to merge it into `angular:compiler-integration`?

File: modules_dart/transform/lib/src/transform/common/code/uri.dart [hunk 1]
```diff
@@ -10,14 +10,13 @@ import 'package:path/path.dart' as path;
 String writeImportUri(String importPath, {String prefix, String fromAbsolute}) {
   var codegenImportPath;
 
-  var resolver = const TransformerUrlResolver();
-  var importUri = resolver.toAssetScheme(Uri.parse(importPath));
+  var importUri = toAssetScheme(Uri.parse(importPath));
```
Comment (yjbanov) [inline modules_dart/transform/lib/src/transform/common/code/uri.dart]: If `importPath` is relative, what does the result of running it through `toAssetScheme` look like?

**Comment by kegluneq** [comment]
Eventually, no, I'll merge into `master`, but setting `angular:compiler-integration` made the diffs correct and not brain melting.

File: modules_dart/transform/lib/src/transform/common/ng_meta.dart [hunk 2]
```diff
@@ -2,74 +2,123 @@ library angular2.transform.common.ng_meta;
...
+/// In the second phase, we perform two actions:
+/// 1. Incorporate all the data from [NgMeta] instances created by all
+///    files `exported` by the original file, such that all aliases and types
+///    visible when importing the original file are stored in its associated
+///    [NgMeta] instance.
+/// 2. Use the [NgDepsModel] to write Dart code registering all injectable
+///    values with the Angular 2 runtime reflection system.
+///
+/// Further down the compilation process, the template compiler needs to reason
+/// about the namespace of import prefixes, so it will combine multple [NgMeta]
```
Comment (yjbanov) [inline modules_dart/transform/lib/src/transform/common/ng_meta.dart]: s/multple/multiple/

File: modules_dart/transform/lib/src/transform/common/xhr_impl.dart [hunk 3]
```diff
@@ -5,22 +5,33 @@ import 'package:angular2/src/core/render/xhr.dart' show XHR;
 import 'package:angular2/src/transform/common/asset_reader.dart';
 import 'package:barback/barback.dart';
 
+/// Transformer-specific implementation of XHR that is backed by an
+/// [AssetReader].
+///
+/// This implementation on supports urls using the asset: scheme.
```
Comment (yjbanov) [inline modules_dart/transform/lib/src/transform/common/xhr_impl.dart]: s/on/only

File: modules_dart/transform/lib/src/transform/common/xhr_impl.dart [hunk 4]
```diff
@@ -5,22 +5,33 @@ import 'package:angular2/src/core/render/xhr.dart' show XHR;
...
-        new AssetId(uri.pathSegments.first, uri.pathSegments.skip(1).join('/'));
+    final assetId = _toAssetId(url);
 
     if (!await _reader.hasInput(assetId)) {
       throw new ArgumentError.value('Could not read asset at uri $url', 'url');
     }
     return _reader.readAsString(assetId);
   }
+
+  Future<bool> exists(String url) => _reader.hasInput(_toAssetId(url));
```
Comment (yjbanov) [inline modules_dart/transform/lib/src/transform/common/xhr_impl.dart]: I wonder if this function should be in `AssetReader`. `XhrImpl` is only there to support angular internals. Transformer code should just use `AssetReader`. No?

File: modules_dart/transform/lib/src/transform/directive_metadata_linker/ng_deps_linker.dart [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+import 'package:angular2/src/core/services/url_resolver.dart';
+import 'package:angular2/src/transform/common/logging.dart';
+import 'package:angular2/src/transform/common/names.dart';
+import 'package:angular2/src/transform/common/model/import_export_model.pb.dart';
+import 'package:angular2/src/transform/common/model/ng_deps_model.pb.dart';
+import 'package:angular2/src/transform/common/url_resolver.dart';
+import 'package:angular2/src/transform/common/xhr_impl.dart';
+
+/// Modifies the [NgDepsModel] represented by `entryPoint` to import its
```
Comment (yjbanov) [inline modules_dart/transform/lib/src/transform/directive_metadata_linker/ng_deps_linker.dart]: s/entryPoint/assetUri/

File: modules_dart/transform/lib/src/transform/common/code/uri.dart (hunk 1 above)
Comment (kegluneq) [inline modules_dart/transform/lib/src/transform/common/code/uri.dart]: An `ArgumentError` :)

Fixed & added lots of tests for `UrlResolver`

**Pull request closed without merging**